src/core/umount.c

```diff
@@ -381,6 +383,112 @@ static bool nonunmountable_path(const char *path) {
                 || path_startswith(path, "/run/initramfs");
 }
 
+static int wait_for_child(pid_t pid) {
+        sigset_t mask;
+        int r, status;
+        usec_t until;
+
+        assert_se(sigemptyset(&mask) == 0);
+        assert_se(sigaddset(&mask, SIGCHLD) == 0);
+
+        /* Drop into a sigtimewait-based timeout. Waiting for the
+         * remount pid to exit. */
+        until = now(CLOCK_MONOTONIC) + DEFAULT_TIMEOUT_USEC;
+        for (;;) {
+                usec_t n;
+                struct timespec ts;
+
+                n = now(CLOCK_MONOTONIC);
+                if (n >= until)
+                        break;
+
+                r = sigtimedwait(&mask, NULL, timespec_store(&ts, until - n));
+                if (r < 0) {
+                        if (errno == EAGAIN) {
+                                /* Timed out, remount child is likely hung,
+                                 * but we can try. */
+                                if (waitpid(pid, &status, WNOHANG) == pid)
+                                        /* Success */
+                                        return (WEXITSTATUS(status) == 0 ? 0 : -WEXITSTATUS(status));
+                                else {
+                                        /* Fail, we should kill the hung
+                                         * process and return */
+                                        log_notice("Timed out, issuing SIGKILL to %d.", pid);
+
+                                        kill(pid, SIGKILL);
```

Comment:
maybe cast this to (void), so that any static checkers know that you ignore this return value on purpose